tests/ui/traits/next-solver/coroutine-higher-ranked-rigid.rs

```diff
@@ -0,0 +1,33 @@
+//@ compile-flags: -Znext-solver
+//@ edition: 2024
+//@ check-pass
+
+// Regression test for <https://github.com/rust-lang/trait-system-refactor-initiative/issues/177>.
+// Coroutines erase all free lifetimes from their interior types. In the future in `from_request`,
+// that means the `'r` lifetime in the type `<T as FromRequest<'r>>::Assoc`, which is in the
+// `dyn Future` type it's awaiting. Normalizing this associated type, with its free lifetimes
+// replaced, means proving `T: FromRequest<'!0>`, which definitely doesn't hold.
+
+// proving `T: Trait` holds when `<T as Trait>::Assoc` is rigid is not necessary for soundness,
+// at least not *yet*, and it's not even necessary for diagnostics since we have other special
+// casing for, e.g., AliasRelate goals failing in the BestObligation folder.
+
+// The old solver deals this by not checking that `T: Trait` holds when `<T as Trait>::Assoc` is
```

Comment:
```suggestion
// The old solver avoids this by not checking that `T: Trait` holds when `<T as Trait>::Assoc` is
```
"deals with" makes it seem like we intentionally decided to not check this which seems not quite true to me.